[OpenMP] atomic compare fail : Codegen support

SunilKuravinakop opened the pull request:
This is a continuation of https://reviews.llvm.org/D123235 ([OpenMP] atomic compare fail : Parser & AST support). In this branch Support for codegen support for atomic compare fail is being added.

Commit:
Adding parameter to fail clause (i.e. memory order clause) for codegen.

  Changes to be committed:
 	modified:   clang/lib/CodeGen/CGStmtOpenMP.cpp

Commit:
Merge branch 'main' into codegen_atomic_compare_fail

Commit:
1) Since the
   {if(x == e) {x = v;} else {d = x;}}
results in generation of cmpxchg in the LLVM IR, the memory order clause
in "fail" clause is now being added.
   In case of other operators ">" & "<" e.g.
   dx = dx > de ? de : dx;
result in atomicrmw in the LLVM IR. "atomicrmw" can have only one memory
order clause. Currently, the memory order clause is ignored.
2) Test cases for codegen
3) In SemaOpenMP.cpp, added
   EncounteredAtomicKinds.contains(OMPC_compare)
instead of AtomicKind == OMPC_compare. This was to accomodate
where AtomicKind is OMPC_capture.

  Changes to be committed:
 	modified:   clang/lib/CodeGen/CGStmtOpenMP.cpp
 	modified:   clang/lib/Sema/SemaOpenMP.cpp
 	modified:   clang/test/OpenMP/atomic_compare_codegen.cpp
 	modified:   llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h
 	modified:   llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp

Commit:
Merge branch 'main' into codegen_atomic_compare_fail

SunilKuravinakop committed:
Merge branch 'llvm:main' into codegen_atomic_compare_fail

alexey-bataev commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -6580,6 +6578,51 @@ void CodeGenFunction::EmitOMPAtomicDirective(const OMPAtomicDirective &S) {
+  if (KindsEncountered.contains(OMPC_compare) &&
+      KindsEncountered.contains(OMPC_fail)) {
+    Kind = OMPC_compare;
+    const OMPFailClause *fC = S.getSingleClause<OMPFailClause>();
```
```suggestion
    const auto *FC = S.getSingleClause<OMPFailClause>();
```

alexey-bataev commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -6580,6 +6578,51 @@ void CodeGenFunction::EmitOMPAtomicDirective(const OMPAtomicDirective &S) {
+    Kind = OMPC_compare;
+    const OMPFailClause *fC = S.getSingleClause<OMPFailClause>();
+    if (fC) {
+      OpenMPClauseKind fP = fC->getFailParameter();
```
```suggestion
      OpenMPClauseKind FP = fC->getFailParameter();
```

ddpagan reviewed (commented):
Aside from Alexey's comments, LGTM.

Commit:
Changes in name format from Alexey.
 Changed the variable names of "fail" clause and the parameter to
this fail clause more self-explanatory.

Commit:
Merge branch 'main' into codegen_atomic_compare_fail

Commit:
Merge branch 'main' into codegen_atomic_compare_fail

Commit:
Merge branch 'main' into codegen_atomic_compare_fail

Commit:
Merge branch 'main' into codegen_atomic_compare_fail
